python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py

```diff
@@ -1213,6 +1214,16 @@ def resolve_tags(self, _graphene_info: ResolveInfo) -> Sequence[GrapheneDefiniti
             if get_definition_tag_type(key) == TagType.USER_PROVIDED
         ]
 
+    def resolve_kinds(self, _graphene_info: ResolveInfo) -> Sequence[str]:
+        kinds = {
+            key.split("/", maxsplit=2)[-1]
+            for key in (self._external_asset_node.tags or {}).keys()
+            if get_definition_tag_type(key) == TagType.HIDDEN and key.startswith(KIND_PREFIX)
+        }
+        if self._external_asset_node.compute_kind:
+            kinds.add(self._external_asset_node.compute_kind)
```

Comment:
I think we should detect this first and early return.

Reply:
In this case you're saying we only return the compute kind, e.g. drop any kinds on the specs? Maybe we should raise an assertion error if there are both kinds on the specs and a compute kind?

Reply:
Either way we should respect one *or* the other, not try to merge them in the graphql layer like this. And as you say, this probably indicates a bug further upstream that we should prevent with assertions. 

And yeah after thinking about it I think we should have the asset kinds dominate, then fall back to compute kind.